src/libraries/System.Text.Json/src/System/Text/Json/Node/JsonObject.List.cs

```diff
@@ -0,0 +1,386 @@
+                }
+
+                _jsonElement = null;
+            }
+
+            _list = list;
+            CreateDictionaryIfThreshold();
+        }
+
+        private int NodeCount
+        {
+            get
+            {
+                CreateList();
+                Debug.Assert(_list != null);
+                return _list.Count;
+            }
+        }
+
+        private JsonNode? SetNode(string propertyName, JsonNode? node)
+        {
+            CreateList();
+            Debug.Assert(_list != null);
+
+            JsonNode? existing = null;
+
+            if (_dictionary != null)
+            {
+                // Fast path if item doesn't exist in dictionary.
+                if (JsonHelpers.TryAdd(_dictionary, propertyName, node))
+                {
+                    node?.AssignParent(this);
+                    CreateDictionaryIfThreshold();
+                    _list.Add(new KeyValuePair<string, JsonNode?>(propertyName, node));
+                    return null;
+                }
+
+                existing = _dictionary[propertyName];
+                if (ReferenceEquals(existing, node))
+                {
+                    _lastKey = propertyName;
+                    _lastValue = node;
+
+                    // Ignore if the same value.
+                    return null;
+                }
+            }
+
+            int i = FindNodeIndex(propertyName);
+            if (i >= 0)
+            {
+                if (_dictionary != null)
+                {
+                    _dictionary[propertyName] = node;
+                }
+                else
+                {
+                    KeyValuePair<string, JsonNode?> current = _list[i];
+                    if (ReferenceEquals(current.Value, node))
+                    {
+                        // Ignore if the same value.
+                        return null;
+                    }
+
+                    existing = current.Value;
+                }
+
+                node?.AssignParent(this);
+                CreateDictionaryIfThreshold();
+                _list[i] = new KeyValuePair<string, JsonNode?>(propertyName, node);
+            }
+            else
+            {
+                node?.AssignParent(this);
+                CreateDictionaryIfThreshold();
+                _dictionary?.Add(propertyName, node);
+                _list.Add(new KeyValuePair<string, JsonNode?>(propertyName, node));
+                Debug.Assert(existing == null);
+            }
+
+            _lastKey = propertyName;
+            _lastValue = node;
+
+            return existing;
+        }
+
+        private int FindNodeIndex(string propertyName)
+        {
+            Debug.Assert(_list != null);
+
+            for (int i = 0; i < _list.Count; i++)
+            {
+                KeyValuePair<string, JsonNode?> current = _list[i];
+                if (GetStringComparer().Compare(propertyName, current.Key) == 0)
```

Comment:
This is performing a full-blown comparison when an equality check should suffice
```suggestion
                if (GetStringComparer().Equals(propertyName, current.Key))
```

Reply:
For a case-insensitive Remove() the Compare is necessary. I'll ensure test coverage for that.

Reply:
Perhaps I'm missing something but this seems to work:
```csharp
> StringComparer.OrdinalIgnoreCase.Equals("abc", "ABC")
true
```

Reply:
The comparer needs to be different for case-sensitive vs. case-insensitive. See the new tests `CaseInsensitive_Remove` and `CaseSensitivity_Remove`.

Reply:
But wouldn't the right comparer be set by the `CreateStringComparer` method regardless?

Reply:
Ah I misread that. I thought you mean `string.Equals()` not `StringComparer.Equals()`. Yes using Equals is better. Thanks.